Add token-based authentication for non-admin users

andrewjjenkins opened the pull request:
Users that do not have the administrator role cannot use basic HTTP authentication for BIG-IP iControl REST.  They must instead get an authentication token that they provide with each request, as described here:
    
https://devcentral.f5.com/wiki/icontrol.authentication_with_the_f5_rest_api.ashx

This pull request adds support for that feature.  It also adds 8 new tests of the functionality.  3 require a non-admin user to be configured on the BIG-IP; they skip if the caller doesn't pass "--nonadmin-username" and "--nonadmin-password".  They pass for me:

```
$ py.test --pdb ./ --bigip=1.2.3.4 --username=admin --password=XXXX
=================================================== test session starts ====================================================
platform darwin -- Python 2.7.8, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /Users/andrew/velcro/f5-icontrol-rest-python, inifile:
plugins: cov-2.2.1
collected 20 items

test_session.py .................sss

========================================== 17 passed, 3 skipped in 12.46 seconds ===========================================
$ py.test --pdb ./ --bigip=1.2.3.4 --username=admin --password=XXXX --nonadmin-username=bob --nonadmin-password=YYYY
=================================================== test session starts ====================================================
platform darwin -- Python 2.7.8, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /Users/andrew/velcro/f5-icontrol-rest-python, inifile:
plugins: cov-2.2.1
collected 20 items

test_session.py ....................

================================================ 20 passed in 17.21 seconds ================================================
```

swormke commented:
The changes look good, but did you happen to build the docs and see if they looked the way you want them?

andrewjjenkins commented:
I did not build the docs.  I'll do that tomorrow.  I should document the
new optional parameter.

On Wed, May 4, 2016, 8:25 PM Shawn Wormke <notifications@github.com> wrote:

> The changes look good, but did you happen to build the docs and see if
> they looked the way you want them?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/F5Networks/f5-icontrol-rest-python/pull/91#issuecomment-217060058>
>

zancas commented on icontrol/session.py:
```diff
@@ -283,6 +254,7 @@ def __init__(self, username, password, **kwargs):
         timeout = kwargs.pop('timeout', 30)
         loglevel = kwargs.pop('loglevel', logging.WARNING)
         logging.getLogger().setLevel(loglevel)
+        token_auth = kwargs.pop('token', None)
```
I prefer `False` for the opposite of `True`.

zancas commented on icontrol/authtoken.py:
```diff
@@ -0,0 +1,137 @@
+        # We don't actually do auth at this point because we don't have a
+        # hostname to authenticate to.
+
+    def token_valid(self):
```
I like methods to be verby, also this method appears to only be called internally, so I propose we change its name to something like:

`_check_token_validity`

I don't really care about the exact name as long as:

It starts with an `_`, and ideally is a verb.   I care more about the `_` part.

zancas commented on icontrol/authtoken.py:
```diff
@@ -0,0 +1,137 @@
+import urlparse
+
+
+class iControlRESTTokenAuth(AuthBase):
```
What's "Auth" short for here?    `Authenticator`, `Authorizer`...   maybe it'll become more obvious to me as I read more.

andrewjjenkins replied:
I'm trying to follow "python requests" here.

https://github.com/kennethreitz/requests/blob/87704105af65b382b86f168f6a54192eab91faf2/requests/auth.py#L71

Everything derives from requests.AuthBase, and is called SomethingAuth.  I'd call one of these requests.Auth instances an Authentication Manager.  Its job in life is to make sure that whenever someone makes a request using requests.Session.post() or similar, it is authenticated using whatever auth technique the Authentication Manager implements.

So for instance, the HTTPBasicAuth class implements the "Basic Auth" in RFC7617.  The iControlRESTTokenAuth implements the token authentication scheme described in the linked devcentral article.

I think there's some value to keeping the ...Auth naming scheme because this is a SomethingAuth class just like everything in requests.auth.  But what comes before, I'm open to suggestions.  I was similarly following iControlRESTSession in session.py.  (I also think Token should be in there somewhere because this is not the most-commonly-documented authentication scheme for BIG-IP).

But really, I'm very much open to suggestions for other names.

zancas replied:
OK, matching the `requests` convention is a persuasive argument, so let's leave it as `iControlRESTTokenAuth`.

I'm not quite sure what's meant by: 'I also thing Token should be in there somewhere....'   what's "there" in this context?   Sorry if I'm being thick here.

zancas commented on icontrol/authtoken.py:
```diff
@@ -0,0 +1,137 @@
+                1000000.0
+            created_bigip = int(respJson['token']['lastUpdateMicros']) / \
+                1000000.0
+        except KeyError:
```
I suppose it's also possible to get a `ValueError` from the `int(FOO)` calls...

zancas commented on icontrol/authtoken.py:
```diff
@@ -0,0 +1,137 @@
+        # the future.  This should account for clock skew between us and
+        # BIG-IP.  By default tokens last for 60 minutes so getting one every
+        # 59 minutes instead of 60 is harmless.
+        if valid_duration > 120.0:
```
What if `valid_duration <= 120.0`?

andrewjjenkins replied:
If the duration is less than 120 seconds, we will attempt to use it for its full duration.  Since we consider the start time to be the time that we first sent the request, that cannot be later than the time the server received the request and created the token.

So, if our clock and BIG-IP's clock are in perfect sync, no issues (and actually in the case of perfect sync, this "renew early" provision is unnecessary).  If our clock is faster than BIG-IP's, also no problem (we'll consider the token expired a little early).

If:
- The BIG-IP is configured to have token durations smaller than 120 seconds, more than 30x smaller than the default.
- Our clock runs slower: it loses X seconds per token duration, where X is the total time it takes us to prepare the request, send it over the network, and for BIG-IP to receive and parse that request and generate a token.
- Clients generate a request right at the end of this token duration.

then we will attempt to use an old token instead of getting a new one, and the request will fail.

zancas replied:
If a failed request is the worst outcome, then I don't think we need additional heuristics.

In the case where the BigIP considers a token expired, and a client does not...   could an attacker (that already knows the victim's password) request a new token, get it, and effectively DOS the victim?   I guess not because it seems that the BigIP will issue a new token even if the requester already holds an unexpired token?

OK, I've talked myself into seeing things your way.

zancas commented on icontrol/session.py:
```diff
@@ -291,9 +263,18 @@ def __init__(self, username, password, **kwargs):
+                                                      token_auth)
+        else:
+            self.session.auth = (username, password)
+
```
Interesting pattern....

zancas commented on icontrol/authtoken.py:
```diff
@@ -0,0 +1,137 @@
+            valid_duration -= 60.0
+        self.expiration = req_start_time + valid_duration
+
+    def __call__(self, r):
```
What's `r`?

andrewjjenkins replied:
request.

Here again I'm following upstream's pattern (https://github.com/kennethreitz/requests/blob/87704105af65b382b86f168f6a54192eab91faf2/requests/auth.py#L215)  but I'll expand.

zancas commented:
Looks great to me mostly.    The only significant concern I have is the case where `valid_duration <= 120.0` really don't know what should happen.

The other stuff is nice-to-have stuff.    Still it's all super simple so hopefully some of it will be worth doing!

About moving exceptions into their own file.   I don't object at all, but I am curious about the motivation.  We can chat about that elsewhere.

andrewjjenkins commented:
OK, I think I addressed your comments in the latest commit @swormke , @zancas .  Please take a look.

Still open:
- Different name for "iControlRESTTokenAuth" or do you buy my argument about following python requests?
- Different way to handle the "token is valid for a very short time, how do we expire early?".  I'm out of ideas and I think the current approach is OK; take a look at my response and LMK if you have better ideas.

zancas commented:
OK, I'm happy with this PR +1 to merge.

zancas commented:
Boom-shaka-laka!
